File: backtracking/797. All Paths From Source to Target/Solution.cs
public class Solution {
    
    private IList<IList<int>> paths;
    
    public IList<IList<int>> AllPathsSourceTarget(int[][] graph) {
        paths = new List<IList<int>>();
        Dfs(graph, 0, graph.Length-1, new List<int>());
        return paths;
    }
    
    private void Dfs(int[][] graph, int currentNode, int targetNode, IList<int> path) {
        
        path.Add(currentNode);
        if (currentNode == targetNode) {
            paths.Add(new List<int>(path));
        }
        else {           
            foreach (int neighbor in graph[currentNode]) {
                Dfs(graph, neighbor, targetNode, path);
            }
        }
        path.RemoveAt(path.Count-1);
    } 
}

File: backtracking/797. All Paths From Source to Target/solution.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        def backtrack(node, path):
            if node == len(graph) - 1:
                result.append(path[:])
                return
            for next_node in graph[node]:
                path.append(next_node)
                backtrack(next_node, path)
                path.pop()
    
        result = []
        backtrack(0, [0])
        return result
